File: android/app/src/test/kotlin/ai/gowtam/oak/features/chat/SlashPickerTest.kt
package ai.gowtam.oak.features.chat

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Slash-discovery P1 lockstep oracle — composer `/` picker model.
 *
 * Clones `web/src/lib/chat/slash-picker.test.ts`. Pure catalog/phase/filter/
 * insert/merge/bind — no I/O, no POST.
 *
 * Expected production API (`features/chat/SlashPicker.kt`):
 *
 *   slashPickerPhase(text):
 *     | SlashPickerPhase.Hidden
 *     | SlashPickerPhase.Commands(prefix, rows)
 *     | SlashPickerPhase.Args(command: "dex" | "team" | "usage", query)
 *     | SlashPickerPhase.Rest(command: "calc" | "new" | "help")
 *
 * Hidden unless the first non-space char is `/`. Empty / whitespace /
 * mid-sentence (`please /dex`) → Hidden (SD-AC-1.4). No space after first
 * token → Commands with that token as prefix (including `/`); `/` lists
 * all six (SD-AC-1.2). `/DEX` with no space stays command phase (SD-AC-3.1).
 * Unknown first token (`/foo`, `/newish`) → Hidden, not an empty command
 * list (SD-AC-1.3) — even though filterCommands("/newish") is [].
 * Space after dex|team|usage → Args. Space after /calc|/new|/help →
 * Rest (no name rows) (SD-BR-6).
 *
 * filterCommands: commandToken.lowercase().startsWith(prefix.lowercase()).
 * prefix "/" → all six. "/newish" → [].
 * insertCommand: trailing space iff catalog trailingSpace.
 * insertName("/dex", "Garchomp") === "/dex Garchomp" (single spaces).
 * mergeDexNameRows: Pokémon, then move, then ability, then item; within a
 * kind keep input order; dedupe kind+slug; cap `limit` default 8 (SD-BR-10,
 * SD-BR-17). Pass already-sliced-per-kind arrays; merge still caps.
 * bindStillValid(bind, composerText): parse is dex navigate AND slashArgs
 * equals bind.displayName (case-insensitive).
 *
 * DexBind / DexNameRow kinds: DexNameKind.Pokemon | Move | Ability | Item.
 *
 * Requirement refs: SD-BR-1, SD-BR-4, SD-BR-5, SD-BR-6, SD-BR-10, SD-BR-17,
 * SD-BR-18, SD-AC-1.2, SD-AC-1.3, SD-AC-1.4, SD-AC-3.1, SD-AC-4.3, SD-AC-7.1.
 */
class SlashPickerTest {

    @Test
    fun `lists exactly the six handled tokens with architecture hints and flags (SD-AC-1_2)`() {
        assertEquals(
            listOf(
                SlashCommandRow(
                    token = "/new",
                    hint = "New empty chat",
                    trailingSpace = false,
                    arg = "none",
                ),
                SlashCommandRow(
                    token = "/team",
                    hint = "Open Teams",
                    hintGuest = "Open Teams · sign in to save",
                    trailingSpace = true,
                    arg = "team",
                ),
                SlashCommandRow(
                    token = "/dex",
                    hint = "Open Dex",
                    trailingSpace = true,
                    arg = "dex",
                ),
                SlashCommandRow(
                    token = "/usage",
                    hint = "Open live usage",
                    trailingSpace = true,
                    arg = "usage",
                ),
                SlashCommandRow(
                    token = "/calc",
                    hint = "Open calculator",
                    trailingSpace = true,
                    arg = "none",
                ),
                SlashCommandRow(
                    token = "/help",
                    hint = "Show these commands",
                    trailingSpace = false,
                    arg = "none",
                ),
            ),
            SLASH_COMMANDS,
        )
        assertEquals(
            listOf("/new", "/team", "/dex", "/usage", "/calc", "/help"),
            SLASH_COMMANDS.map { it.token },
        )
    }

    @Test
    fun `exports the picker caption and arg-phase empty lines`() {
        assertEquals("Insert, then send", PICKER_CAPTION)
        assertEquals("No Dex matches", EMPTY_DEX)
        assertEquals("No usage matches", EMPTY_USAGE)
        assertEquals("No saved teams match", EMPTY_TEAMS)
        assertEquals("Sign in to save teams", EMPTY_TEAMS_GUEST)
    }

    @Test
    fun `keeps commands whose token starts with the prefix, case-insensitive (SD-AC-1_2, SD-BR-5)`() {
        assertEquals(
            listOf("/new", "/team", "/dex", "/usage", "/calc", "/help"),
            filterCommands("/").map { it.token },
        )
        assertEquals(SLASH_COMMANDS, filterCommands("/"))
        assertEquals(listOf("/dex"), filterCommands("/de").map { it.token })
        assertEquals(listOf("/dex"), filterCommands("/DEX").map { it.token })
        assertEquals(listOf("/new"), filterCommands("/n").map { it.token })
        assertEquals(listOf("/new"), filterCommands("/new").map { it.token })
        assertEquals(listOf("/help"), filterCommands("/h").map { it.token })
    }

    @Test
    fun `returns no rows for slash-newish — slash-new does not start with slash-newish (SD-AC-1_2, SD-AC-1_3)`() {
        assertEquals(emptyList<SlashCommandRow>(), filterCommands("/newish"))
        assertEquals(emptyList<SlashCommandRow>(), filterCommands("/foo"))
    }

    @Test
    fun `is hidden when the first non-space char is not slash (SD-AC-1_4)`() {
        assertEquals(SlashPickerPhase.Hidden, slashPickerPhase(""))
        assertEquals(SlashPickerPhase.Hidden, slashPickerPhase("   "))
        assertEquals(SlashPickerPhase.Hidden, slashPickerPhase("please /dex"))
        assertEquals(SlashPickerPhase.Hidden, slashPickerPhase("what about /team later"))
    }

    @Test
    fun `opens commands for a leading slash prefix and lists all six (SD-AC-1_2)`() {
        assertEquals(
            SlashPickerPhase.Commands(prefix = "/", rows = SLASH_COMMANDS),
            slashPickerPhase("/"),
        )
        assertEquals(
            SlashPickerPhase.Commands(prefix = "/", rows = SLASH_COMMANDS),
            slashPickerPhase("  /"),
        )
    }

    @Test
    fun `filters command rows by the first token while there is no space (SD-AC-1_2, SD-AC-3_1)`() {
        assertEquals(
            SlashPickerPhase.Commands(
                prefix = "/de",
                rows = SLASH_COMMANDS.filter { it.token == "/dex" },
            ),
            slashPickerPhase("/de"),
        )
        assertEquals(
            SlashPickerPhase.Commands(
                prefix = "/DEX",
                rows = SLASH_COMMANDS.filter { it.token == "/dex" },
            ),
            slashPickerPhase("/DEX"),
        )
        assertEquals(
            SlashPickerPhase.Commands(
                prefix = "/dex",
                rows = SLASH_COMMANDS.filter { it.token == "/dex" },
            ),
            slashPickerPhase("/dex"),
        )
        assertEquals(
            SlashPickerPhase.Commands(
                prefix = "/calc",
                rows = SLASH_COMMANDS.filter { it.token == "/calc" },
            ),
            slashPickerPhase("/calc"),
        )
        assertEquals(
            SlashPickerPhase.Commands(
                prefix = "/new",
                rows = SLASH_COMMANDS.filter { it.token == "/new" },
            ),
            slashPickerPhase("/new"),
        )
        assertEquals(
            SlashPickerPhase.Commands(
                prefix = "/help",
                rows = SLASH_COMMANDS.filter { it.token == "/help" },
            ),
            slashPickerPhase("/help"),
        )
    }

    @Test
    fun `hides for an unknown first token, not an empty command list (SD-AC-1_3)`() {
        assertEquals(SlashPickerPhase.Hidden, slashPickerPhase("/newish"))
        assertEquals(SlashPickerPhase.Hidden, slashPickerPhase("/foo"))
        assertEquals(emptyList<SlashCommandRow>(), filterCommands("/newish"))
    }

    @Test
    fun `enters args after a space on slash-dex slash-team slash-usage (SD-BR-6)`() {
        assertEquals(
            SlashPickerPhase.Args(command = "dex", query = ""),
            slashPickerPhase("/dex "),
        )
        assertEquals(
            SlashPickerPhase.Args(command = "dex", query = "gar"),
            slashPickerPhase("/dex gar"),
        )
        assertEquals(
            SlashPickerPhase.Args(command = "dex", query = "gar"),
            slashPickerPhase("  /dex gar"),
        )
        assertEquals(
            SlashPickerPhase.Args(command = "dex", query = "Garchomp"),
            slashPickerPhase("/DEX Garchomp"),
        )
        assertEquals(
            SlashPickerPhase.Args(command = "team", query = ""),
            slashPickerPhase("/team "),
        )
        assertEquals(
            SlashPickerPhase.Args(command = "team", query = "Rain Offense"),
            slashPickerPhase("/team Rain Offense"),
        )
        assertEquals(
            SlashPickerPhase.Args(command = "usage", query = ""),
            slashPickerPhase("/usage "),
        )
        assertEquals(
            SlashPickerPhase.Args(command = "usage", query = "garchomp"),
            slashPickerPhase("/usage garchomp"),
        )
    }

    @Test
    fun `enters rest after a space on slash-calc slash-new slash-help — no name rows (SD-BR-6)`() {
        assertEquals(SlashPickerPhase.Rest(command = "calc"), slashPickerPhase("/calc "))
        assertEquals(SlashPickerPhase.Rest(command = "calc"), slashPickerPhase("/calc foo vs bar"))
        assertEquals(SlashPickerPhase.Rest(command = "new"), slashPickerPhase("/new "))
        assertEquals(SlashPickerPhase.Rest(command = "new"), slashPickerPhase("/new rain team"))
        assertEquals(SlashPickerPhase.Rest(command = "help"), slashPickerPhase("/help "))
        assertEquals(SlashPickerPhase.Rest(command = "help"), slashPickerPhase("/help extra words"))
        val calcRest = slashPickerPhase("/calc foo")
        assertTrue(calcRest is SlashPickerPhase.Rest)
        assertFalse(calcRest is SlashPickerPhase.Commands)
        assertFalse(calcRest is SlashPickerPhase.Args)
    }

    @Test
    fun `appends a trailing space iff the catalog trailingSpace flag is true (SD-AC-2_1, SD-AC-2_2)`() {
        assertEquals("/dex ", insertCommand("/dex"))
        assertEquals("/team ", insertCommand("/team"))
        assertEquals("/usage ", insertCommand("/usage"))
        assertEquals("/calc ", insertCommand("/calc"))
        assertEquals("/new", insertCommand("/new"))
        assertEquals("/help", insertCommand("/help"))
    }

    @Test
    fun `joins command and display name with a single space and no required trailing space (SD-AC-2_3)`() {
        assertEquals("/dex Garchomp", insertName("/dex", "Garchomp"))
        assertEquals("/usage Garchomp", insertName("/usage", "Garchomp"))
        assertEquals("/team Rain Offense", insertName("/team", "Rain Offense"))
    }

    @Test
    fun `orders Pokémon, then move, then ability, then item — keeps within-kind input order (SD-BR-17)`() {
        val item = DexNameRow(
            kind = DexNameKind.Item,
            slug = "metronome",
            displayName = "Metronome",
        )
        val ability = DexNameRow(
            kind = DexNameKind.Ability,
            slug = "rough-skin",
            displayName = "Rough Skin",
        )
        val move = DexNameRow(
            kind = DexNameKind.Move,
            slug = "earthquake",
            displayName = "Earthquake",
        )
        val zamazenta = DexNameRow(
            kind = DexNameKind.Pokemon,
            slug = "zamazenta",
            displayName = "Zamazenta",
        )
        val garchomp = DexNameRow(
            kind = DexNameKind.Pokemon,
            slug = "garchomp",
            displayName = "Garchomp",
            spriteUrl = "https://example.com/garchomp.png",
        )

        assertEquals(
            listOf(zamazenta, garchomp, move, ability, item),
            mergeDexNameRows(
                listOf(
                    DexKindMatches(kind = DexNameKind.Item, matches = listOf(item)),
                    DexKindMatches(kind = DexNameKind.Ability, matches = listOf(ability)),
                    DexKindMatches(kind = DexNameKind.Move, matches = listOf(move)),
                    DexKindMatches(
                        kind = DexNameKind.Pokemon,
                        matches = listOf(zamazenta, garchomp),
                    ),
                ),
            ),
        )
    }

    @Test
    fun `dedupes by kind+slug, keeping the first occurrence`() {
        val first = DexNameRow(
            kind = DexNameKind.Pokemon,
            slug = "garchomp",
            displayName = "Garchomp",
        )
        val duplicate = DexNameRow(
            kind = DexNameKind.Pokemon,
            slug = "garchomp",
            displayName = "GARCHOMP",
        )
        val metronomeMove = DexNameRow(
            kind = DexNameKind.Move,
            slug = "metronome",
            displayName = "Metronome",
        )
        val metronomeItem = DexNameRow(
            kind = DexNameKind.Item,
            slug = "metronome",
            displayName = "Metronome",
        )

        assertEquals(
            listOf(first, metronomeMove, metronomeItem),
            mergeDexNameRows(
                listOf(
                    DexKindMatches(
                        kind = DexNameKind.Pokemon,
                        matches = listOf(first, duplicate),
                    ),
                    DexKindMatches(kind = DexNameKind.Move, matches = listOf(metronomeMove)),
                    DexKindMatches(kind = DexNameKind.Item, matches = listOf(metronomeItem)),
                ),
            ),
        )
    }

    @Test
    fun `caps at limit default 8, preserving merge order (SD-BR-10)`() {
        val pokemon = (0 until 9).map { i ->
            DexNameRow(
                kind = DexNameKind.Pokemon,
                slug = "p$i",
                displayName = "P$i",
            )
        }

        val capped = mergeDexNameRows(
            listOf(DexKindMatches(kind = DexNameKind.Pokemon, matches = pokemon)),
        )
        assertEquals(8, capped.size)
        assertEquals(
            listOf("p0", "p1", "p2", "p3", "p4", "p5", "p6", "p7"),
            capped.map { it.slug },
        )
    }

    @Test
    fun `still caps after already-sliced-per-kind arrays (SD-BR-10)`() {
        fun sliced(kind: DexNameKind) = (0 until 8).map { i ->
            DexNameRow(
                kind = kind,
                slug = "${kind.name.lowercase()}-$i",
                displayName = "${kind.name.lowercase()} $i",
            )
        }

        val merged = mergeDexNameRows(
            listOf(
                DexKindMatches(kind = DexNameKind.Pokemon, matches = sliced(DexNameKind.Pokemon)),
                DexKindMatches(kind = DexNameKind.Move, matches = sliced(DexNameKind.Move)),
                DexKindMatches(kind = DexNameKind.Ability, matches = sliced(DexNameKind.Ability)),
                DexKindMatches(kind = DexNameKind.Item, matches = sliced(DexNameKind.Item)),
            ),
        )

        assertEquals(8, merged.size)
        assertTrue(merged.all { it.kind == DexNameKind.Pokemon })
        assertEquals(
            listOf(
                "pokemon-0",
                "pokemon-1",
                "pokemon-2",
                "pokemon-3",
                "pokemon-4",
                "pokemon-5",
                "pokemon-6",
                "pokemon-7",
            ),
            merged.map { it.slug },
        )
    }

    @Test
    fun `bindStillValid is true when parse is dex navigate and slashArgs equals displayName (SD-BR-17)`() {
        val garchomp = DexBind(
            kind = DexNameKind.Pokemon,
            slug = "garchomp",
            displayName = "Garchomp",
        )
        val metronomeMove = DexBind(
            kind = DexNameKind.Move,
            slug = "metronome",
            displayName = "Metronome",
        )

        assertTrue(bindStillValid(garchomp, "/dex Garchomp"))
        assertTrue(bindStillValid(garchomp, "/DEX garchomp"))
        assertTrue(bindStillValid(garchomp, "  /dex GARCHOMP"))
        assertTrue(bindStillValid(metronomeMove, "/dex Metronome"))
    }

    @Test
    fun `bindStillValid is false when the name is edited, the composer is not dex, or the command changes (SD-BR-17)`() {
        val garchomp = DexBind(
            kind = DexNameKind.Pokemon,
            slug = "garchomp",
            displayName = "Garchomp",
        )

        assertFalse(bindStillValid(garchomp, "/dex Garchom"))
        assertFalse(bindStillValid(garchomp, "/dex GarchompX"))
        assertFalse(bindStillValid(garchomp, "/dex"))
        assertFalse(bindStillValid(garchomp, "/team Garchomp"))
        assertFalse(bindStillValid(garchomp, "/usage Garchomp"))
        assertFalse(bindStillValid(garchomp, "/calc Garchomp"))
        assertFalse(bindStillValid(garchomp, "/new"))
        assertFalse(bindStillValid(garchomp, "/dexish Garchomp"))
        assertFalse(bindStillValid(garchomp, "please /dex Garchomp"))
    }
}

File: web/src/lib/chat/slash-picker.test.ts
/**
 * Slash-discovery P1 lockstep oracle — composer `/` picker model.
 *
 * iOS `SlashPickerTests` and Android `SlashPickerTest` must clone these
 * cases. Pure catalog/phase/filter/insert/merge/bind — no I/O, no POST.
 *
 * Requirement refs: SD-BR-1, SD-BR-4, SD-BR-5, SD-BR-6, SD-BR-10, SD-BR-17,
 * SD-BR-18, SD-AC-1.2, SD-AC-1.3, SD-AC-1.4, SD-AC-3.1, SD-AC-4.3, SD-AC-7.1.
 *
 *   slashPickerPhase(text):
 *     | { phase: "hidden" }
 *     | { phase: "commands"; prefix: string; rows: CommandRow[] }
 *     | { phase: "args"; command: "dex" | "team" | "usage" | "calc"; query: string }
 *     | { phase: "rest"; command: "new" | "help" }
 *
 * Hidden unless the first non-space char is `/`. Empty / whitespace /
 * mid-sentence (`please /dex`) → hidden (SD-AC-1.4). No space after first
 * token → `commands` with that token as prefix (including `/`); `/` lists
 * all six (SD-AC-1.2). `/DEX` with no space stays command phase (SD-AC-3.1).
 * Unknown first token (`/foo`, `/newish`) → hidden, not an empty command
 * list (SD-AC-1.3) — even though filterCommands("/newish") is [].
 * Space after dex|team|usage|calc → `args` (SD-BR-6, SD-US-10). Space
 * after /new|/help → `rest` (no name rows).
 *
 * filterCommands: commandToken.toLowerCase().startsWith(prefix.toLowerCase()).
 * prefix "/" → all six. "/newish" → [].
 * insertCommand: trailing space iff catalog trailingSpace.
 * insertName("/dex", "Garchomp") === "/dex Garchomp" (single spaces).
 * mergeDexNameRows: Pokémon, then move, then ability, then item; within a
 * kind keep input order; dedupe kind+slug; cap `limit` default 8 (SD-BR-10,
 * SD-BR-17). Pass already-sliced-per-kind arrays; merge still caps.
 * bindStillValid(bind, composerText): parse is dex navigate AND slashArg
 * equals bind.displayName (case-insensitive).
 *
 * DexBind / DexNameRow kinds: "pokemon" | "move" | "ability" | "item".
 */

import { describe, expect, it } from "vitest";

import {
  EMPTY_DEX,
  EMPTY_TEAMS,
  EMPTY_TEAMS_GUEST,
  EMPTY_USAGE,
  PICKER_CAPTION,
  SLASH_COMMANDS,
  bindStillValid,
  filterCommands,
  insertCommand,
  insertName,
  mergeDexNameRows,
  slashPickerPhase,
} from "./slash-picker";

describe("SLASH_COMMANDS catalog", () => {
  it("lists exactly the six handled tokens with architecture hints and flags (SD-AC-1.2)", () => {
    expect(SLASH_COMMANDS).toEqual([
      {
        token: "/new",
        hint: "New empty chat",
        trailingSpace: false,
        arg: "none",
      },
      {
        token: "/team",
        hint: "Open Teams",
        hintGuest: "Open Teams · sign in to save",
        trailingSpace: true,
        arg: "team",
      },
      {
        token: "/dex",
        hint: "Open Dex",
        trailingSpace: true,
        arg: "dex",
      },
      {
        token: "/usage",
        hint: "Open live usage",
        trailingSpace: true,
        arg: "usage",
      },
      {
        token: "/calc",
        hint: "Open calculator",
        trailingSpace: true,
        arg: "calc",
      },
      {
        token: "/help",
        hint: "Show these commands",
        trailingSpace: false,
        arg: "none",
      },
    ]);
    expect(SLASH_COMMANDS.map((row) => row.token)).toEqual([
      "/new",
      "/team",
      "/dex",
      "/usage",
      "/calc",
      "/help",
    ]);
  });
});

describe("picker copy constants", () => {
  it("exports the picker caption and arg-phase empty lines", () => {
    expect(PICKER_CAPTION).toBe("Insert, then send");
    expect(EMPTY_DEX).toBe("No Dex matches");
    expect(EMPTY_USAGE).toBe("No usage matches");
    expect(EMPTY_TEAMS).toBe("No saved teams match");
    expect(EMPTY_TEAMS_GUEST).toBe("Sign in to save teams");
  });
});

describe("filterCommands", () => {
  it("keeps commands whose token starts with the prefix, case-insensitive (SD-AC-1.2, SD-BR-5)", () => {
    expect(filterCommands("/").map((row) => row.token)).toEqual([
      "/new",
      "/team",
      "/dex",
      "/usage",
      "/calc",
      "/help",
    ]);
    expect(filterCommands("/")).toEqual([...SLASH_COMMANDS]);
    expect(filterCommands("/de").map((row) => row.token)).toEqual(["/dex"]);
    expect(filterCommands("/DEX").map((row) => row.token)).toEqual(["/dex"]);
    expect(filterCommands("/n").map((row) => row.token)).toEqual(["/new"]);
    expect(filterCommands("/new").map((row) => row.token)).toEqual(["/new"]);
    expect(filterCommands("/h").map((row) => row.token)).toEqual(["/help"]);
  });

  it("returns no rows for /newish — /new does not start with /newish (SD-AC-1.2, SD-AC-1.3)", () => {
    expect(filterCommands("/newish")).toEqual([]);
    expect(filterCommands("/foo")).toEqual([]);
  });
});

describe("slashPickerPhase", () => {
  it("is hidden when the first non-space char is not / (SD-AC-1.4)", () => {
    expect(slashPickerPhase("")).toEqual({ phase: "hidden" });
    expect(slashPickerPhase("   ")).toEqual({ phase: "hidden" });
    expect(slashPickerPhase("please /dex")).toEqual({ phase: "hidden" });
    expect(slashPickerPhase("what about /team later")).toEqual({
      phase: "hidden",
    });
  });

  it("opens commands for a leading / prefix and lists all six (SD-AC-1.2)", () => {
    expect(slashPickerPhase("/")).toEqual({
      phase: "commands",
      prefix: "/",
      rows: [...SLASH_COMMANDS],
    });
    expect(slashPickerPhase("  /")).toEqual({
      phase: "commands",
      prefix: "/",
      rows: [...SLASH_COMMANDS],
    });
  });

  it("filters command rows by the first token while there is no space (SD-AC-1.2, SD-AC-3.1)", () => {
    expect(slashPickerPhase("/de")).toEqual({
      phase: "commands",
      prefix: "/de",
      rows: SLASH_COMMANDS.filter((row) => row.token === "/dex"),
    });
    expect(slashPickerPhase("/DEX")).toEqual({
      phase: "commands",
      prefix: "/DEX",
      rows: SLASH_COMMANDS.filter((row) => row.token === "/dex"),
    });
    expect(slashPickerPhase("/dex")).toEqual({
      phase: "commands",
      prefix: "/dex",
      rows: SLASH_COMMANDS.filter((row) => row.token === "/dex"),
    });
    expect(slashPickerPhase("/calc")).toEqual({
      phase: "commands",
      prefix: "/calc",
      rows: SLASH_COMMANDS.filter((row) => row.token === "/calc"),
    });
    expect(slashPickerPhase("/new")).toEqual({
      phase: "commands",
      prefix: "/new",
      rows: SLASH_COMMANDS.filter((row) => row.token === "/new"),
    });
    expect(slashPickerPhase("/help")).toEqual({
      phase: "commands",
      prefix: "/help",
      rows: SLASH_COMMANDS.filter((row) => row.token === "/help"),
    });
  });

  it("hides for an unknown first token, not an empty command list (SD-AC-1.3)", () => {
    expect(slashPickerPhase("/newish")).toEqual({ phase: "hidden" });
    expect(slashPickerPhase("/foo")).toEqual({ phase: "hidden" });
    expect(filterCommands("/newish")).toEqual([]);
  });

  it("enters args after a space on /dex /team /usage /calc (SD-BR-6, SD-US-10)", () => {
    expect(slashPickerPhase("/dex ")).toEqual({
      phase: "args",
      command: "dex",
      query: "",
    });
    expect(slashPickerPhase("/dex gar")).toEqual({
      phase: "args",
      command: "dex",
      query: "gar",
    });
    expect(slashPickerPhase("  /dex gar")).toEqual({
      phase: "args",
      command: "dex",
      query: "gar",
    });
    expect(slashPickerPhase("/DEX Garchomp")).toEqual({
      phase: "args",
      command: "dex",
      query: "Garchomp",
    });
    expect(slashPickerPhase("/team ")).toEqual({
      phase: "args",
      command: "team",
      query: "",
    });
    expect(slashPickerPhase("/team Rain Offense")).toEqual({
      phase: "args",
      command: "team",
      query: "Rain Offense",
    });
    expect(slashPickerPhase("/usage ")).toEqual({
      phase: "args",
      command: "usage",
      query: "",
    });
    expect(slashPickerPhase("/usage garchomp")).toEqual({
      phase: "args",
      command: "usage",
      query: "garchomp",
    });
    expect(slashPickerPhase("/calc ")).toEqual({
      phase: "args",
      command: "calc",
      query: "",
    });
    expect(slashPickerPhase("/calc foo vs bar")).toEqual({
      phase: "args",
      command: "calc",
      query: "foo vs bar",
    });
    expect(slashPickerPhase("/CALC Garchomp Earthquake vs Gholdengo")).toEqual({
      phase: "args",
      command: "calc",
      query: "Garchomp Earthquake vs Gholdengo",
    });
  });

  it("enters rest after a space on /new /help — no name rows (SD-BR-6)", () => {
    expect(slashPickerPhase("/new ")).toEqual({
      phase: "rest",
      command: "new",
    });
    expect(slashPickerPhase("/new rain team")).toEqual({
      phase: "rest",
      command: "new",
    });
    expect(slashPickerPhase("/help ")).toEqual({
      phase: "rest",
      command: "help",
    });
    expect(slashPickerPhase("/help extra words")).toEqual({
      phase: "rest",
      command: "help",
    });
    expect(slashPickerPhase("/calc foo")).not.toHaveProperty("rows");
  });
});

describe("insertCommand", () => {
  it("appends a trailing space iff the catalog trailingSpace flag is true (SD-AC-2.1, SD-AC-2.2)", () => {
    expect(insertCommand("/dex")).toBe("/dex ");
    expect(insertCommand("/team")).toBe("/team ");
    expect(insertCommand("/usage")).toBe("/usage ");
    expect(insertCommand("/calc")).toBe("/calc ");
    expect(insertCommand("/new")).toBe("/new");
    expect(insertCommand("/help")).toBe("/help");
  });
});

describe("insertName", () => {
  it("joins command and display name with a single space and no required trailing space (SD-AC-2.3)", () => {
    expect(insertName("/dex", "Garchomp")).toBe("/dex Garchomp");
    expect(insertName("/usage", "Garchomp")).toBe("/usage Garchomp");
    expect(insertName("/team", "Rain Offense")).toBe("/team Rain Offense");
  });
});

describe("mergeDexNameRows", () => {
  it("orders Pokémon, then move, then ability, then item; keeps within-kind input order (SD-BR-17)", () => {
    const item = {
      kind: "item" as const,
      slug: "metronome",
      displayName: "Metronome",
    };
    const ability = {
      kind: "ability" as const,
      slug: "rough-skin",
      displayName: "Rough Skin",
    };
    const move = {
      kind: "move" as const,
      slug: "earthquake",
      displayName: "Earthquake",
    };
    const zamazenta = {
      kind: "pokemon" as const,
      slug: "zamazenta",
      displayName: "Zamazenta",
    };
    const garchomp = {
      kind: "pokemon" as const,
      slug: "garchomp",
      displayName: "Garchomp",
      spriteUrl: "https://example.com/garchomp.png",
    };

    expect(
      mergeDexNameRows([
        { kind: "item", matches: [item] },
        { kind: "ability", matches: [ability] },
        { kind: "move", matches: [move] },
        { kind: "pokemon", matches: [zamazenta, garchomp] },
      ]),
    ).toEqual([zamazenta, garchomp, move, ability, item]);
  });

  it("dedupes by kind+slug, keeping the first occurrence", () => {
    const first = {
      kind: "pokemon" as const,
      slug: "garchomp",
      displayName: "Garchomp",
    };
    const duplicate = {
      kind: "pokemon" as const,
      slug: "garchomp",
      displayName: "GARCHOMP",
    };
    const metronomeMove = {
      kind: "move" as const,
      slug: "metronome",
      displayName: "Metronome",
    };
    const metronomeItem = {
      kind: "item" as const,
      slug: "metronome",
      displayName: "Metronome",
    };

    expect(
      mergeDexNameRows([
        { kind: "pokemon", matches: [first, duplicate] },
        { kind: "move", matches: [metronomeMove] },
        { kind: "item", matches: [metronomeItem] },
      ]),
    ).toEqual([first, metronomeMove, metronomeItem]);
  });

  it("caps at limit default 8, preserving merge order (SD-BR-10)", () => {
    const pokemon = Array.from({ length: 9 }, (_, i) => ({
      kind: "pokemon" as const,
      slug: `p${i}`,
      displayName: `P${i}`,
    }));

    const capped = mergeDexNameRows([{ kind: "pokemon", matches: pokemon }]);
    expect(capped).toHaveLength(8);
    expect(capped.map((row) => row.slug)).toEqual([
      "p0",
      "p1",
      "p2",
      "p3",
      "p4",
      "p5",
      "p6",
      "p7",
    ]);
  });

  it("still caps after already-sliced-per-kind arrays (SD-BR-10)", () => {
    const sliced = (kind: "pokemon" | "move" | "ability" | "item") =>
      Array.from({ length: 8 }, (_, i) => ({
        kind,
        slug: `${kind}-${i}`,
        displayName: `${kind} ${i}`,
      }));

    const merged = mergeDexNameRows([
      { kind: "pokemon", matches: sliced("pokemon") },
      { kind: "move", matches: sliced("move") },
      { kind: "ability", matches: sliced("ability") },
      { kind: "item", matches: sliced("item") },
    ]);

    expect(merged).toHaveLength(8);
    expect(merged.every((row) => row.kind === "pokemon")).toBe(true);
    expect(merged.map((row) => row.slug)).toEqual([
      "pokemon-0",
      "pokemon-1",
      "pokemon-2",
      "pokemon-3",
      "pokemon-4",
      "pokemon-5",
      "pokemon-6",
      "pokemon-7",
    ]);
  });
});

describe("bindStillValid", () => {
  const garchomp = {
    kind: "pokemon" as const,
    slug: "garchomp",
    displayName: "Garchomp",
  };
  const metronomeMove = {
    kind: "move" as const,
    slug: "metronome",
    displayName: "Metronome",
  };

  it("is true when parse is dex navigate and slashArg equals displayName (SD-BR-17)", () => {
    expect(bindStillValid(garchomp, "/dex Garchomp")).toBe(true);
    expect(bindStillValid(garchomp, "/DEX garchomp")).toBe(true);
    expect(bindStillValid(garchomp, "  /dex GARCHOMP")).toBe(true);
    expect(bindStillValid(metronomeMove, "/dex Metronome")).toBe(true);
  });

  it("is false when the name is edited, the composer is not dex, or the command changes (SD-BR-17)", () => {
    expect(bindStillValid(garchomp, "/dex Garchom")).toBe(false);
    expect(bindStillValid(garchomp, "/dex GarchompX")).toBe(false);
    expect(bindStillValid(garchomp, "/dex")).toBe(false);
    expect(bindStillValid(garchomp, "/team Garchomp")).toBe(false);
    expect(bindStillValid(garchomp, "/usage Garchomp")).toBe(false);
    expect(bindStillValid(garchomp, "/calc Garchomp")).toBe(false);
    expect(bindStillValid(garchomp, "/new")).toBe(false);
    expect(bindStillValid(garchomp, "/dexish Garchomp")).toBe(false);
    expect(bindStillValid(garchomp, "please /dex Garchomp")).toBe(false);
  });
});
